Kernel/Bus/USB/EHCI/EHCIController.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023, Leon Albrecht <leon.a@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Bus/PCI/API.h>
+#include <Kernel/Bus/USB/EHCI/EHCIController.h>
+
+namespace Kernel::USB::EHCI {
+
+ErrorOr<NonnullLockRefPtr<EHCIController>> EHCIController::try_to_initialize(const PCI::DeviceIdentifier& pci_device_identifier)
+{
+
+    // FIXME: This assumes the BIOS left us a physical region for the controller
+    u64 pci_bar_value = PCI::get_BAR(pci_device_identifier, SpaceBaseAddressRegister);
+    auto pci_bar_space_type = PCI::get_BAR_space_type(pci_bar_value);
+    if (pci_bar_space_type == PCI::BARSpaceType::Memory64BitSpace) {
+        u64 next_pci_bar_value = PCI::get_BAR(pci_device_identifier, static_cast<PCI::HeaderType0BaseRegister>(to_underlying(SpaceBaseAddressRegister) + 1));
+        pci_bar_value |= next_pci_bar_value << 32;
+    }
+    auto pci_bar_space_size = PCI::get_BAR_space_size(pci_device_identifier, SpaceBaseAddressRegister);
+    auto register_map = TRY(Memory::map_typed<void>(PhysicalAddress(pci_bar_value & PCI::bar_address_mask), pci_bar_space_size, Memory::Region::Access::ReadWrite));
+
+    PCI::enable_bus_mastering(pci_device_identifier);
+    PCI::enable_memory_space(pci_device_identifier);
+
+    auto controller = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) EHCIController(pci_device_identifier, move(register_map))));
+
+    TRY(controller->initialize());
+
+    return controller;
+}
+
+EHCIController::EHCIController(PCI::DeviceIdentifier const& pci_device_identifier, Memory::TypedMapping<void> mapped_registers)
+    : PCI::Device(pci_device_identifier)
+    , m_mapped_registers(move(mapped_registers))
+{
+    m_cap_regs = bit_cast<CapabilityRegisters const*>(m_mapped_registers.base_address().get());
+    m_op_regs = bit_cast<OperationalRegisters volatile*>(m_mapped_registers.base_address().get() + m_cap_regs->capability_length);
```

Comment:
Hm, like, can we map this set of regs beforehand? who is pulling random values from m_mapped_registers later, not through the cap/op regs?